Wikipedia/Code/WMFURLSchemeHandler.m

```diff
@@ -1,439 +0,0 @@
-
-        if ([imageTag isSizeLargeEnoughForGalleryInclusion]) {
-            resizedSrc = [[imageTag URLForTargetWidth:targetImageWidth] absoluteString];
-            if (resizedSrc) {
-                src = resizedSrc;
-            }
-        }
-
-        if (src) {
-            NSString *srcWithProxy = [NSURL wmf_appSchemeURLForURLString:src].absoluteString;
-            if (srcWithProxy) {
-                NSString *newSrcAttribute = [@[@"src=\"", srcWithProxy, @"\""] componentsJoinedByString:@""];
-                imageTag.src = newSrcAttribute;
-                newImageTagContents = [imageTag.imageTagContents mutableCopy];
-            }
-        }
-    }
-
-    [newImageTagContents replaceOccurrencesOfString:@"srcset" withString:@"data-srcset-disabled" options:0 range:NSMakeRange(0, newImageTagContents.length)]; //disable the srcset since we put the correct resolution image in the src
-
-    if (resizedSrc) {
-        [newImageTagContents appendString:@" data-image-gallery=\"true\""]; //the javascript looks for this to know if it should attempt widening
-    }
-
-    return newImageTagContents;
-}
-
-#pragma mark - Cache
-
-- (void)setResponseData:(nullable NSData *)data withContentType:(nullable NSString *)contentType forPath:(NSString *)path requestURL:(NSURL *)requestURL {
-    NSMutableDictionary<NSString *, NSString *> *headerFields = [NSMutableDictionary dictionaryWithCapacity:1];
-    if (contentType) {
-        headerFields[@"Content-Type"] = contentType;
-    }
-    NSHTTPURLResponse *response = [[NSHTTPURLResponse alloc] initWithURL:requestURL statusCode:200 HTTPVersion:nil headerFields:headerFields];
-    [self cacheResponse:response data:data forPath:path];
-}
-
-- (void)cacheResponse:(NSURLResponse *)response data:(NSData *)data forPath:(NSString *)path {
-    if (path == nil) {
-        return;
-    }
-    NSCachedURLResponse *cachedResponse = [[NSCachedURLResponse alloc] initWithResponse:response data:data];
-    [self.responseCache setObject:cachedResponse forKey:path];
-}
-
-- (NSCachedURLResponse *)cachedResponseForPath:(NSString *)path {
-    if (path == nil) {
-        return nil;
-    }
-    return [self.responseCache objectForKey:path];
-}
-
-- (void)cacheSectionDataForArticle:(MWKArticle *)article {
-    NSString *articleKey = article.url.wmf_articleDatabaseKey;
-    if (articleKey == nil) {
-        return;
-    }
-    [self.articleCache setObject:article forKey:articleKey];
-}
-
-- (MWKArticle *)articleForKey:(NSString *)path {
-    if (path == nil) {
-        return nil;
-    }
-    return [self.articleCache objectForKey:path];
-}
-
-#pragma mark - BaseURL (for testing only)
-
-- (NSURL *)baseURL {
-    return [[self baseURLComponents] URL];
-}
-
-- (void)webView:(nonnull WKWebView *)webView startURLSchemeTask:(nonnull id<WKURLSchemeTask>)urlSchemeTask {
-    WMFAssertMainThread(@"startURLSchemeTask assumed to be called on the main thread");
-
-    [self.activeTasks addObject:urlSchemeTask];
```

Comment:
SchemeHandler from the lite project keeps track of the URLSessionTasks returned from Session, but WMFURLSchemeHandler kept track of the URLSchemeTasks. I didn't keep URLSchemeTask tracking when changing to Swift but let me know if that should be preserved.

Reply:
As long as you can guarantee that no other calls are made on the `WKURLSchemeTask` after `stop urlSchemeTask` is called, it will be fine. Without tracking the `WKURLSchemeTask` you'll need to check for the `URLSessionTask` in the `Session.Callback` blocks. Also, we discovered in https://github.com/wikimedia/wikipedia-ios/pull/2982 that it looks like WebKit is ensuring thread safety on `WKURLSchemeTask` by using the main thread, so there will need to be `DspatchQueue.main.async`s inside of the callback blocks. At that point `tasks` should be access controlled by the main thread as well, eliminating the need for the private `SchemeHandlerQueue`